### PR TITLE
Allow multiple public inputs in verify api

### PR DESCRIPTION
**Pull request by jonas-lj**
The current api for the Groth16 verifier implicitly expected only a single public input. This PR fixes this.

### CODE REVIEW

File: fastcrypto-zkp/src/api.rs [hunk 1]
```diff
@@ -31,8 +33,13 @@ pub fn verify_groth16_in_bytes(
     proof_public_inputs_as_bytes: &[u8],
     proof_points_as_bytes: &[u8],
 ) -> Result<bool, FastCryptoError> {
-    let x = BlsFr::deserialize_compressed(proof_public_inputs_as_bytes)
-        .map_err(|_| FastCryptoError::InvalidInput)?;
+    let mut x = Vec::new();
+    for chunk in proof_public_inputs_as_bytes.chunks(SCALAR_LENGTH) {
```
Comment (kchalkias) [inline fastcrypto-zkp/src/api.rs:37]: Can we document that we get a concatenated bytes, then split the bytes into SCALAR_LENGTH chuncks?

File: fastcrypto-zkp/src/api.rs [hunk 2]
```diff
@@ -31,8 +33,13 @@ pub fn verify_groth16_in_bytes(
...
+        if chunk.len() != SCALAR_LENGTH {
```
Comment (kchalkias) [inline fastcrypto-zkp/src/api.rs:38]: you can check earlier if the `proof_public_inputs_as_bytes.len mod SCALAR_LENGTH` = 0 before parsing anything, to fail fast

File: fastcrypto-zkp/src/api.rs [hunk 3]
```diff
@@ -31,8 +33,13 @@ pub fn verify_groth16_in_bytes(
     proof_public_inputs_as_bytes: &[u8],
     proof_points_as_bytes: &[u8],
 ) -> Result<bool, FastCryptoError> {
-    let x = BlsFr::deserialize_compressed(proof_public_inputs_as_bytes)
-        .map_err(|_| FastCryptoError::InvalidInput)?;
+    let mut x = Vec::new();
```
Comment (kchalkias) [inline fastcrypto-zkp/src/api.rs:36]: We could also return fast if it's an empty public inputs. We won't fail, its possible for a circuit to not have public inputs, we need a test for that too.

**Review by kchalkias — APPROVED** [review]
Approving, but we need to follow with upcoming PRs:
a) test against NO public inputs (empty), @benr-ml wdyt?
b) @benr-ml what about the reported memory leak from auditor?

File: fastcrypto-zkp/src/api.rs [hunk 4]
```diff
@@ -44,6 +51,5 @@ pub fn verify_groth16_in_bytes(
         delta_g2_neg_pc_bytes,
     )?;
 
-    verify_with_processed_vk(&blst_pvk, &[x], &proof)
-        .map_err(|_| FastCryptoError::GeneralOpaqueError)
+    verify_with_processed_vk(&blst_pvk, &x, &proof).map_err(|_| FastCryptoError::GeneralOpaqueError)
```
Comment (benr-ml) [inline fastcrypto-zkp/src/api.rs:54]: since there is nothing secret here, we should & can use a more explicit error (e.g., GeneralError(msg))

**Comment by benr-ml** [comment]
> Approving, but we need to follow with upcoming PRs: a) test against NO public inputs (empty), @benr-ml wdyt? b) @benr-ml what about the reported memory leak from auditor?

Yes and yes, both should be done

**Comment by jonas-lj** [comment]
> > Approving, but we need to follow with upcoming PRs: a) test against NO public inputs (empty), @benr-ml wdyt? b) @benr-ml what about the reported memory leak from auditor?
> 
> 
> 
> Yes and yes, both should be done

I've added a new issue.

File: fastcrypto-zkp/src/api.rs (hunk 4 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/api.rs:54]: That's fixed in #1623
